src/core_plugins/elasticsearch/lib/cluster.js

```diff
@@ -12,89 +11,104 @@ export default class Cluster {
-  callAsKibanaUserFactory(client) {
-    return partial(this.callWithRequestFactory(client), undefined);
+  callWithRequest = (req = {}, endpoint, clientParams = {}, options = {}) => {
+    if (req.headers) {
+      const filteredHeaders = filterHeaders(req.headers, this.config('requestHeadersWhitelist'));
+      set(clientParams, 'headers', filteredHeaders);
+    }
+
+    return callAPI(this._noAuthClient, endpoint, clientParams, options);
   }
 
-  callWithRequestFactory(client) {
-    /**
-     * callWithRequest
-     *
-     * Makes a call to ES, passing through whitelisted headers in the request
-     *
-     * The whitelisted headers are defined in the config under _requestHeadersWhitelist_
-     *
-     * @param {Object|undefined} req - The request object
-     * @param {string} endpoint
-     * @param {Object} clientParams
-     * @param {Object} options
-     * @param {boolean} options.wrap401Errors
-     * @returns {Promise}
-     */
-    return (req = {}, endpoint, clientParams = {}, options = {}) => {
-      const wrap401Errors = options.wrap401Errors !== false;
-
-      if (req.headers) {
-        const filteredHeaders = filterHeaders(req.headers, this.config('requestHeadersWhitelist'));
-        set(clientParams, 'headers', filteredHeaders);
-      }
-
-      const clientPath = toPath(endpoint);
-      const api = get(client, clientPath);
-
-      let apiContext = get(client, clientPath.slice(0, -1));
-      if (isEmpty(apiContext)) {
-        apiContext = client;
-      }
-
-      if (!api) {
-        throw new Error(`called with an invalid endpoint: ${endpoint}`);
-      }
-
-      return api.call(apiContext, clientParams).catch((err) => {
-        if (!wrap401Errors || err.statusCode !== 401) {
-          return Promise.reject(err);
-        }
-
-        const boomError = Boom.wrap(err, err.statusCode);
-        const wwwAuthHeader = get(err, 'body.error.header[WWW-Authenticate]');
-        boomError.output.headers['WWW-Authenticate'] = wwwAuthHeader || 'Basic realm="Authorization Required"';
-
-        throw boomError;
-      });
-    };
+  /**
+   * callAsKibanaUser
+   *
+   * Performs a call to ES using the credentials in kibana.yml
+   *
+   * @param {string} endpoint
+   * @param {Object} clientParams
+   * @param {Object} options
+   * @returns {Promise}
+   */
+
+  callAsKibanaUser = (endpoint, clientParams = {}, options = {}) => {
+    return callAPI(this._client, endpoint, clientParams, options);
   }
 
-  config(path) {
+  config = (path) => {
     return cloneDeep(path ? get(this._config, path) : this._config);
   }
 
-  createClient(options = {}) {
-    return createClient(Object.assign({}, this.config(), options));
+  addClientPlugins(plugins = []) {
+    this.close(); // close existing client connections
+
+    if (Array.isArray(this._config.plugins)) {
+      this._config.plugins = this._config.plugins.concat(plugins);
+    } else {
+      this._config.plugins = plugins;
+    }
+
+    createClients.bind(this)();
```

Comment:
Pushed a fix. I played around previously with passing config to createClients and having it return an array or object containing the clients but it was messier than I would have liked.